### ISSUE TITLE
Raku regex fails to perform the Perl equivalent one 

### ISSUE THREAD

**Issue by abdulbadii** [rakudo/rakudo#4764]
## The Problem
Raku regex equivalent to translate the Perl one fails, as it's been translated 

## Expected Behavior
Must behave the Perl one

## Actual Behavior
It's hanging
doesn't succeed nor fail on which return 

## Steps to Reproduce
This 
https://github.com/abdulbadii/HTML-element-operations/blob/main/rmHTMLement.pl

has been tried to get translated to Raku

https://github.com/abdulbadii/HTML-XML-Operation-Raku

In regex of below sub fails at above hash characters

```
sub getAllDepth ($tag, $nth, $search, @ret, $OFF, $min, $nrev) {
	my (@nd, $offset,	$offs);
	my @curNode= ([$OFF, $search],);
	while @curNode {
		for @curNode -> $OffNode {
			if $nrev {
				my $i=1;	$OffNode[1] ~~ /^<HEAD> [ <C> [ \<$tag>>.+ & <ND> ] { ++$i } ]+/;	
				return 1 if ($nth=$i-$nth) < 1}
			$OffNode[1] ~~ /^(<HEAD>) { $offset=$0 }
			[ [ (<A>) { $offs=($offset~=$1) }
			[ <!before \<$tag>> > (<N>)
			{ push @nd, ($OffNode[0]~$offset, $2) if $MAX > $min;	$offset~=$2; } ]? ]*
\###########################################
			[ \<$tag>>.+ & (<N>) ]
			{ push @nd, ($OffNode[0]~$offs, $3) if $MAX > $min; $offset=$offs~$3 }
			] **{$nth}
			{ push @ret, ($OffNode[0]~$offs, $3) if $MAX >= $min }
			[ (<A>) { $offset~=$4 }
			(<N>) { push @nd, ($OffNode[0]~$offset, $5) if $MAX > $min; $offset~=$5 } ]* /
		}
		@curNode=@nd; @nd=();
	}
	return !@ret
}
```

This is the original Perl's
```
sub getAllDepth {	# $_[1] nth/nth bckwrd  $_[2] search space element											# $_[4] its offset  $_[5] depth  $_[6] nth bckwrd
	my ($ret, $min, $onref, @nd,$offset,$offs) = ($_[3], $_[5]);
	my @curNode=[$_[4], $_[2]];
	while (@curNode) {
		for $onref (@curNode) {
			if ($_[6]) {
				my $c=1; $onref->[1]=~ /^$HEAD(?:$C(?=<$_[0]\b)$ND(?{ ++$c }))+/;	$_[1]=$c-$_[6]}
			$onref->[1]=~
			/^($HEAD) (?{ $offset=$1 }) (?:
			(?: ($A) (?{$offs=$offset.=$2 })
			(?: (?!<$_[0]\b) (?'n'$N)
			(?{ push (@nd, [$onref->[0].$offset, $+{n}]) if $MAX>$min;
			$offset.=$+{n} })
			)? )*+
			(?=<$_[0]\b) (?'nd'$N)
			(?{ push (@nd, [$onref->[0].$offs, $+{nd}]) if $MAX>$min;
			$offset=$offs.$+{nd} })
			) {$_[1]}
			(?{ push (@$ret, [$onref->[0].$offs, $+{nd}]) if $MAX>=$min })
			(?: (?'a'$A) (?{ $offset.=$+{a} })
			(?'z'$N) (?{ push (@nd, [$onref->[0].$offset, $+{z}]) if $MAX>$min; $offset.=$+{z} })
			)*/x
		}
		@curNode=@nd; @nd=();
	}
	return !@$ret
}

```## Environment
* Operating system: MoarVM on Arch Linux
*
* Compiler version (`perl6 -v` or `raku -v`): Rakudo™ v2021.12.
Implementing the Raku® Programming Language v6.d

**Comment by vrurg** [comment]
Please, golf down your sample to some short kind of code which would be easier to read and understand. It would take too much time to read and understand your example and people barely have it, unfortunately. We're a small community.

Otherwise, I'd recommend studying grammars as they're seemingly suit your task better. In either way, directly translating Perl regexes into Raku is not a very good idea. They're different, especially when it comes to some edge cases.

**Comment by abdulbadii** [comment]
Thanks GOD
to guide me and help find the boiled-down issue 

```

> 'foo' ~~ / [l?]+ /
^CPressed CTRL-c, press CTRL-c again to exit

> 'foo' ~~ / [l?]* /
^CPressed CTRL-c, press CTRL-c again to exit

```
it's hanging
so must break it disruptly 

while
```
 DB<97> 'foo' =~ / (?: l? )* /x

  DB<98> 'foo' =~ / (?: l? )+ /x

  DB<99> 

```
 Perl seamlessly works

**Comment by tbrowder** [comment]
What are you trying to do?

**Comment by 0racle** [comment]
> What are you trying to do?

We could argue the merits of what the user is doing, but a more productive question is: what _should_ this code do. 

The regex is trying to match **0 or more** repetitions of **0 or 1** patterns. Perl presumably progresses the cursor (or maybe just optimises this away) but Raku hangs. Another way to test this issue is with the following snippet: -
```raku
'foo' ~~ / [ 'l'? ] ** 250_000 /
```

On my machine, this hangs for about a second before giving up. Without an upper limit (as is the case with the `+` and `*` quantifiers) it just hangs indefinitely. I also tried this pattern with `:ratchet` and it made no difference.

(This snippet is also a potential interim work-around for the user, but perhaps they want to lower the rep count to something that makes sense for their data, so that it doesn't cause too much of a slow down.)

That said, Raku Regexes are no Perl regexes, and we don't guarantee that they will semantically do the same thing with similar patterns. So again the question is... is what Raku does here logical/intended, or is it an unintended bug?

**Comment by 0racle** [comment]
It looks like the cursor does progress to the end of the string, but keeps trying to match.

```
> 'foo' ~~ / [ 'l'? ] ** 100 <!{ say .pos => .prematch.raku with $/ }> /
0 => ""
1 => "f"
2 => "fo"
3 => "foo"
```

**Comment by jnthn** [comment]
> is what Raku does here logical/intended, or is it an unintended bug?

The policy so far as been no more than: if it's a direct atom that trivially can never consume characters, then produce a compile-time error (for example `$+`). A group could consume characters, so no compile-time error. There's no runtime "did we make progress" checks. Having them would magically "fix" rules that contain infinite loops, but given rules are relatively late-bound, we'd pay for it all over the place; it would also not be much fun for those using the engine for its backtracking to drive non-string-matching logic, I guess.

**Comment by pmichaud** [comment]
This subject (infinite repetition of zero-width matches) comes up a lot.  I'm a little surprised it's not mentioned in https://docs.raku.org/language/regexes-best-practices , or anywhere on docs.raku.org as far as I can tell.  Should it be...?

**Comment by jnthn** [comment]
> This subject (infinite repetition of zero-width matches) comes up a lot. 

Yes, enough that I sometimes wonder if it would be better to insist that a `+` or `*` quantifier checks for progress, although knowing it would also slow down parsing of Raku source gives some pause.

> Should it be...?

Yes.

**Comment by pmichaud** [comment]
> Yes, enough that I sometimes wonder if it would be better to insist that a `+` or `*` quantifier checks for progress, 
> although knowing it would also slow down parsing of Raku source gives some pause.

I often try to think of ways that checking for progress could be done efficiently.  The other possibility -- someone wants to have the zero-width repetition for its side effects -- seems like a de-optimization, flag, or special syntax that could be made available for that case.

Overall I've tended to think of it as "when repeating, see if we advanced in the last repetition", which feels expensive.   Tonight I'm wondering if quantifiers could be forward looking -- i.e., when starting a new quant iteration, set a marker that says "in order to repeat this quantifier again, the cursor needs to be at least at this position....".  In the usual case that position would simply be "current position + 1", but some exotic things could happen, including using current position + 0 to mean allow quanitification of zero-width results.

Anyway, just some thoughts here.  The standard at the moment is that Raku is not guaranteed to terminate when quantifying a zero-width expression, and that's how it has been... well.... since the beginning.

Pm
